### PR TITLE
misc(contrib) : add contribution guidelines

### PR DESCRIPTION
**Pull request by amarouni**
#713 First PR to add a draft contribution rules and commit message formatting. I also added a PR template.

### CODE REVIEW

**Comment by amarouni** [comment]
@noorul @velvia Did you get sometime to review this one ?

**Comment by velvia** [comment]
I’ll have a look later today

**Comment by noorul** [comment]
@amarouni I agree that we need some guidelines and process. But shouldn't we first implement them in the code base and then publish it to new contributors?

**Comment by velvia** [comment]
@amarouni thanks for this. I like this template and agree with most of it....  I think... maybe we start informally enforcing this and use it for our own PRs, then get a feel for what works and doesn't.

@noorul I don't think anyone will bother to go back and fix all the thousands of commits etc....  it's probably only practical to implement going forwards. Besides revising git history is somewhat fragile and dangerous.

**Comment by amarouni** [comment]
@velvia Ok sounds good to me, let's start with the currently open PRs then see what works and what deson't then we can formalize it.

**Comment by amarouni** [comment]
@noorul @velvia So what do you think about this now ?

**Comment by velvia** [comment]
I’m OK with it so far.

**Comment by noorul** [comment]
@amarouni I have responded by some review comments

**Comment by amarouni** [comment]
@noorul I don't see any comments, are you sure that you submitted the review ?

**Comment by noorul** [comment]
@amarouni I did not. Damn workflow change, https://xkcd.com/1172/

File: doc/contribution-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Contribution guidelines
+
+- [Pull Requests](#pullrequest)
+- [Commit Message Guidelines](#commit)
+
+## <a name="pullrequest"></a> Pull Requests Guidelines
+Pull requests are the only means by which you can contribute to this project, please follow the following steps when submitting pull requests :
+
+1. Create your pull request. (PR name must match [commit name](#commit))
```
Comment (amarouni) [inline doc/contribution-guidelines.md]: The first step, creating the PR from the users fork with the PR name respecting the commit format name.

File: doc/contribution-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+2. Fill the pull request template with all of the required elements
+3. CI tests kick in and report back any issues, you should fix these issues before continuing. No broken PR will be reviewed.
```
Comment (amarouni) [inline doc/contribution-guidelines.md]: I agree with you, these were inspired from private work where developers have sufficient knowledge to debug failing PRs. We can get rid of it

File: doc/contribution-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+#### Type
+Must be one of the following:
+
+* **feat**: A new feature
+* **fix**: A bug fix
+* **refactor**: Refactoring change
+* **docs**: Documentation only changes
+* **style**: Style changes (white-space, formatting, missing semi-colons, etc)
+* **test**: Adding missing tests
+* **chore**: Changes to the build process or auxiliary tools and libraries
```
Comment (amarouni) [inline doc/contribution-guidelines.md]: Don't mind. I'll change it

File: doc/contribution-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+#### Scope
+The scope sets the scope of the commit, for example `webapi`, `akka-app`, `doa`, etc...
+
+#### Subject
+Brief commit description
+
+* use the imperative, present tense: "change" not "changed" nor "changes"
+* don't capitalize first letter
```
Comment (amarouni) [inline doc/contribution-guidelines.md]: it's just an aesthetic thing, it keeps the whole commit message in a consistent case. And from my personal experience it makes the change log easier to read !

**Comment by amarouni** [comment]
@noorul Yes and I updated the PR to include your feedback.

File: doc/contribution-guidelines.md [hunk 5]
```diff
@@ -94,4 +94,4 @@ fix(webapi): fix webapi timeout
 
 ```
 chore(ci): add new CI tests
```
Comment (noorul) [inline doc/contribution-guidelines.md]: Shouldn't this be misc?

**Comment by amarouni** [comment]
@noorul Yes my bad I've fixed it.

**Comment by amarouni** [comment]
@noorul I merged master on my branch and update the PR title. Once everything is approved we can squash merge the whole PR (and that's what we should be doing from now on, no more PR merge without squash) this will leave a single commit in the correct format.

**Comment by amarouni** [comment]
@noorul @velvia good to go ?

**Comment by noorul** [comment]
@amarouni Yes, please squash

**Comment by amarouni** [comment]
@noorul When merging the PR you can choose **merge squash** and it will automatically squash all of the commits into a single one and then merge it with master. That's what I was trying ti say in my previous comment. That's what I think the merging workflow should look like, the user simply commits to the branch and the person merging the PR should select the squash option. Since not all users are git wizards this can make easier for all the users.
What do you think ?

**Comment by noorul** [comment]
@amarouni I think the author should squash and push that we he can edit the commit log accordingly.

**Comment by velvia** [comment]
@amarouni I did not know that you could squash and merge.  Let me try that.
@noorul I'm OK with squashing if it is easy.  If there are merge conflicts the author should do that.

**Comment by noorul** [comment]
@velvia Most of the time you might have to edit the commit log. I would better leave it to the author.

**Comment by velvia** [comment]
I just replaced the default commit log (which has a list of all commits) with the title of the PR.  I do agree in general it might be better for authors to do it - maybe they can include the description they want.  Or maybe the PR title can be the default

-Evan
"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead
"We can do no great things, just small things with great love." -mother Teresa

**Pull request merged**